feat(finance): add ISO 4217 numerical codes to Currency object

Nfloc opened the pull request:
Address issue #2289 

added numericCode parameter to currency object in all languages (put zero for all non ISO currency (cypto currency)).

added currencyNumericCode to finance definition.

added currencyNumericCode testing to finance.spec.ts

changed descriptions to match new parameter.

added random currencyNumericCode method.

Nfloc committed:
feat(finance): Added ISO 4217 numerical codes to Currency object

added numID parameter to currency object in all languages (put zero for all non ISO currency (cypto currency)).

added currencyNumID to finance definition.

added currencyNumID testing to finance.spec.ts

changed descriptions to match new parameter.

added random currencyNumID method

Nfloc committed:
Merge branch 'next' into next

matthewmayer commented on src/modules/finance/index.ts:
```diff
@@ -516,6 +522,19 @@ export class FinanceModule extends ModuleBase {
+   * @example
+   * faker.finance.currencyNumID() // 840
+   *
+   * @since 9.4.0
```
Next release will be 9.6.0

matthewmayer commented on src/modules/finance/index.ts:
```diff
@@ -516,6 +522,19 @@ export class FinanceModule extends ModuleBase {
+   * (The ISO 4217 numerical code for the currency (e.g. `US Dollar` -> 840 ))
+   *
+   * @example
+   * faker.finance.currencyNumID() // 840
```
Don't like numID as a name

Suggest

numericCode
numericalCode
number

matthewmayer replied:
https://fakerjs.dev/api/location#countrycode uses 'numeric' so i think `numericCode`/`currencyNumericCode()` would probably be most consistent?

ST-DDT commented on src/locales/fa/finance/currency.ts:
```diff
@@ -3,135 +3,162 @@ export default [
     name: 'درهم امارات',
     code: 'AED',
     symbol: '',
+    numID: 784, // UAE Dirham
```
Sometimes you have a comment here and sometimes you dont.
Would you mind sharing your intend for the comment or is it just a dev artifact?

Nfloc replied:
Mostly a dev artifact. 
For all the non-english languages I put the English translations so it was easier to read and compare.

edit: all translation comments removed

ST-DDT commented on src/modules/finance/index.ts:
```diff
@@ -516,6 +522,19 @@ export class FinanceModule extends ModuleBase {
+   *
+   * @since 9.4.0
+   */
+  currencyNumID(): number {
```
I'm not sure we need a new method as `faker.finance.currency().numID` is easy enough.
What do you think?

Shinigami92 replied:
When you have a look into the code above this function implementation, you will find that we currently also have functions for the other `currency*` functions. IMO we should at least be consistent, so I'm fine with having that new extra function.

matthewmayer replied:
i agree for consistency with other methods lets keep the extra method.

matthewmayer commented on src/modules/finance/index.ts:
```diff
@@ -516,6 +522,19 @@ export class FinanceModule extends ModuleBase {
+   *
+   * @since 9.4.0
+   */
+  currencyNumID(): number {
```
Should it actually return a string rather than a number? It appears the codes are usually presented as three digits with leading zeros, e.g. AUD is `036` not `36`?

https://en.wikipedia.org/wiki/ISO_4217#cite_note-SOS217-1

<img width="853" alt="image" src="https://github.com/user-attachments/assets/e2d7ac16-045f-4149-a29d-d7c0c060e292" />

Nfloc replied:
Good question, in the original posted issue the wording indicating that it would be a number instead of a string.


- ![419140da62a0f21134aed2ff18cb6c3f](https://github.com/user-attachments/assets/871023d8-4357-414c-b3df-5a2fb0b9053c)

Let me know opinions on what it should be, I'm not super familiar with financial applications.

Nfloc committed:
feat(finance): numericCode module parameter

Changed previous parameter numID --> numericCode.
Changed comments to reflect change and newest release.

Nfloc committed:
feat(finance): el numericCode

Changed parameter numID --> numericCode

Nfloc committed:
feat(finance): en numericCode

Changed parameter numID --> numericCode

Nfloc committed:
feat(finance): fa numericCode

Changed parameter numID --> numericCode.
Removed comment dev artifacts.

Nfloc committed:
feat(finance): fr numericCode

Changed parameter numID --> numericCode.
Removed dev artifact comments

Nfloc committed:
feat(finance): zh_CN numericCode

Changed parameter numID --> numericCode.
Removed dev artifact comments.

Nfloc committed:
feat(finance): update definition comments

Changed comment from numID to numericCode

Nfloc committed:
test(finance): numericCode naming

Updated to new numericCode naming (numID previous)

Nfloc committed:
test(finance): numericCode naming

Changed parameter numID --> numericCode
Changed method currencyNumID --> currencyNumericCode

Nfloc committed:
test(finance): numericCode naming

Changed method currencyNumID --> currencyNumericCode

ST-DDT commented on src/modules/finance/index.ts:
```diff
@@ -457,14 +462,15 @@ export class FinanceModule extends ModuleBase {
    * @see faker.finance.currencyCode(): For generating specifically the currency code.
    * @see faker.finance.currencyName(): For generating specifically the currency name.
    * @see faker.finance.currencySymbol(): For generating specifically the currency symbol.
+   * @see faker.finance.numericCode(): For generating specifically the currency numerical ID.
```
```suggestion
   * @see faker.finance.currencyNumericCode(): For generating specifically the currency numerical ID.
```

ST-DDT commented on src/modules/finance/index.ts:
```diff
@@ -27,6 +27,11 @@ export interface Currency {
+  /**
+   * The ISO 4217 numeric code for the currency (e.g. `840`).
+   */
+  numericCode: number;
```
Since the result is always 3 characters long and doesn't need the usual number operations like addition, we should use a string here.

Nfloc committed:
Merge pull request #1 from faker-js/next

chore(deps): lock file maintenance (#3403)

Nfloc committed:
feat(finance): change numericCode to string

finance object parameter numericCode has been updated to a string.
numericCode has been given significant zeros. (e.g. '8' -> '008')

Nfloc committed:
Merge pull request #2 from faker-js/next

fix: test before using Buffers (#3400)

Nfloc commented:
Should I update my original pull request description to fit the new parameter names?

ST-DDT commented on src/locales/fa/finance/currency.ts:
```diff
@@ -1,137 +1,932 @@
 export default [
   {
-    name: 'درهم امارات',
+    name: '阿联酋迪拉姆',
```
I think you unintentionally mixed the names up in this file.

ST-DDT commented on src/modules/finance/index.ts:
```diff
@@ -516,6 +522,19 @@ export class FinanceModule extends ModuleBase {
 
+  /**
+   * Returns a random currency number ID.
+   * (The ISO 4217 numerical code for the currency (e.g. `US Dollar` -> '840' ))
```
```suggestion
   * (The ISO 4217 numerical code for the currency (e.g. `US Dollar` -> `'840'` ))
```

ST-DDT commented on src/modules/finance/index.ts:
```diff
@@ -516,6 +522,19 @@ export class FinanceModule extends ModuleBase {
   }
 
+  /**
+   * Returns a random currency number ID.
```
```suggestion
   * Returns a random currency numeric code.
```

ST-DDT commented on src/modules/finance/index.ts:
```diff
@@ -457,14 +462,15 @@ export class FinanceModule extends ModuleBase {
    * @see faker.finance.currencyCode(): For generating specifically the currency code.
    * @see faker.finance.currencyName(): For generating specifically the currency name.
    * @see faker.finance.currencySymbol(): For generating specifically the currency symbol.
+   * @see faker.finance.currencyNumericCode(): For generating specifically the currency numerical ID.
```
```suggestion
   * @see faker.finance.currencyNumericCode(): For generating specifically the currency numeric code.
```

Nfloc committed:
feat(finance): update src module description

add grave markers around numerical code description.

Co-authored-by: ST-DDT <ST-DDT@gmx.de>

Nfloc committed:
feat(finance): update src module description

change numerical ID to numeric code in description.

Co-authored-by: ST-DDT <ST-DDT@gmx.de>

Nfloc committed:
feat(finance): update src module description

change number ID to numeric code in description.

Co-authored-by: ST-DDT <ST-DDT@gmx.de>

Nfloc committed:
feat(finance): language revert change

fa language file was overridden by zh_CHN langauge file, changes reverted back to original.

matthewmayer commented on test/modules/finance.spec.ts:
```diff
@@ -326,6 +328,14 @@ describe('finance', () => {
+        it('should return a string', () => {
+          const currencyNumericCode = faker.finance.currencyNumericCode();
+
+          expect(currencyNumericCode).toBeTypeOf('string');
```
also check its 3 characters long?

Nfloc committed:
test(finance): add length testcase

add test case to ensure that currencyNumericCode method returns a string exactly 3 digits long.

xDivisionByZerox commented on src/modules/finance/index.ts:
```diff
@@ -516,6 +522,19 @@ export class FinanceModule extends ModuleBase {
 
+  /**
+   * Returns a random currency numeric code.
+   * (The ISO 4217 numerical code for the currency (e.g. `US Dollar` -> `'840'` ))
```
Are the parenthesis required here? Additionally, since this function returns a random code, it should be "a currency" not "the curreny":


```suggestion
   * The ISO 4217 numerical code for a currency (e.g. `US Dollar` -> `'840'` )
```

xDivisionByZerox commented on src/definitions/finance.ts:
```diff
@@ -18,7 +18,7 @@ export type FinanceDefinition = LocaleEntry<{
 
   /**
-   * Currencies including their name, code and symbol (e.g. `US Dollar` / `USD` / `$`).
+   * Currencies including their name, code, symbol and ISO numericCode (e.g. `US Dollar` / `USD` / `$` / '840').
```
This should be written in plain English and not a reference to the property name:


```suggestion
   * Currencies including their name, code, symbol and ISO numerical code (e.g. `US Dollar` / `USD` / `$` / '840').
```

Nfloc committed:
feat(finance): Update finance.ts

update grammar in currency object descriptor.

Nfloc committed:
feat(finance): Update index.ts

update grammar and syntax in currencyNumericCode(); descriptor

ST-DDT committed:
Merge branch 'next' into next

ST-DDT committed:
Merge branch 'next' into next

ST-DDT committed:
Merge branch 'next' into next

ST-DDT commented:
Thanks for your contribution ❤️
